src/webgpu/shader/validation/types/matrix.spec.ts

```diff
@@ -0,0 +1,141 @@
+
+const kValidCases = {
+  // Basic matrices
+  mat2x2_f32: `alias T = mat2x2<f32>;`,
+  mat2x3_f32: `alias T = mat2x3<f32>;`,
+  mat2x4_f32: `alias T = mat2x4<f32>;`,
+  mat3x2_f32: `alias T = mat3x2<f32>;`,
+  mat3x3_f32: `alias T = mat3x3<f32>;`,
+  mat3x4_f32: `alias T = mat3x4<f32>;`,
+  mat4x2_f32: `alias T = mat4x2<f32>;`,
+  mat4x3_f32: `alias T = mat4x3<f32>;`,
+  mat4x4_f32: `alias T = mat4x4<f32>;`,
+  mat2x2_f16: `enable f16;\nalias T = mat2x2<f16>;`,
+  mat2x3_f16: `enable f16;\nalias T = mat2x3<f16>;`,
+  mat2x4_f16: `enable f16;\nalias T = mat2x4<f16>;`,
+  mat3x2_f16: `enable f16;\nalias T = mat3x2<f16>;`,
+  mat3x3_f16: `enable f16;\nalias T = mat3x3<f16>;`,
+  mat3x4_f16: `enable f16;\nalias T = mat3x4<f16>;`,
+  mat4x2_f16: `enable f16;\nalias T = mat4x2<f16>;`,
+  mat4x3_f16: `enable f16;\nalias T = mat4x3<f16>;`,
+  mat4x4_f16: `enable f16;\nalias T = mat4x4<f16>;`,
+
+  // Pre-declared aliases
+  mat2x2f: `alias T = mat2x2f;`,
+  mat2x3f: `alias T = mat2x3f;`,
+  mat2x4f: `alias T = mat2x4f;`,
+  mat3x2f: `alias T = mat3x2f;`,
+  mat3x3f: `alias T = mat3x3f;`,
+  mat3x4f: `alias T = mat3x4f;`,
+  mat4x2f: `alias T = mat4x2f;`,
+  mat4x3f: `alias T = mat4x3f;`,
+  mat4x4f: `alias T = mat4x4f;`,
+  mat2x2h: `enable f16;\nalias T = mat2x2h;`,
+  mat2x3h: `enable f16;\nalias T = mat2x3h;`,
+  mat2x4h: `enable f16;\nalias T = mat2x4h;`,
+  mat3x2h: `enable f16;\nalias T = mat3x2h;`,
+  mat3x3h: `enable f16;\nalias T = mat3x3h;`,
+  mat3x4h: `enable f16;\nalias T = mat3x4h;`,
+  mat4x2h: `enable f16;\nalias T = mat4x2h;`,
+  mat4x3h: `enable f16;\nalias T = mat4x3h;`,
+  mat4x4h: `enable f16;\nalias T = mat4x4h;`,
+
+  trailing_comman: `alias T = mat2x2<f32,>;`,
+
+  // Abstract matrices
+  abstract_2x2: `const m = mat2x2(1,1,1,1);`,
+  abstract_2x3: `const m = mat2x3(1,1,1,1,1,1);`,
+  abstract_2x4: `const m = mat2x4(1,1,1,1,1,1,1,1);`,
+
+  // Pre-declared shadowable
+  shadow_mat2x2f: `alias mat2x2f = mat2x2<f32>;`,
+  shadow_mat2x3f: `alias mat2x3f = mat2x3<f32>;`,
+  shadow_mat2x4f: `alias mat2x4f = mat2x4<f32>;`,
+  shadow_mat3x2f: `alias mat3x2f = mat3x2<f32>;`,
+  shadow_mat3x3f: `alias mat3x3f = mat3x3<f32>;`,
+  shadow_mat3x4f: `alias mat3x4f = mat3x4<f32>;`,
+  shadow_mat4x2f: `alias mat4x2f = mat4x2<f32>;`,
+  shadow_mat4x3f: `alias mat4x3f = mat4x3<f32>;`,
+  shadow_mat4x4f: `alias mat4x4f = mat4x4<f32>;`,
+  shadow_mat2x2h: `enable f16;\nalias mat2x2h = mat2x2<f16>;`,
+  shadow_mat2x3h: `enable f16;\nalias mat2x3h = mat2x3<f16>;`,
+  shadow_mat2x4h: `enable f16;\nalias mat2x4h = mat2x4<f16>;`,
+  shadow_mat3x2h: `enable f16;\nalias mat3x2h = mat3x2<f16>;`,
+  shadow_mat3x3h: `enable f16;\nalias mat3x3h = mat3x3<f16>;`,
+  shadow_mat3x4h: `enable f16;\nalias mat3x4h = mat3x4<f16>;`,
+  shadow_mat4x2h: `enable f16;\nalias mat4x2h = mat4x2<f16>;`,
+  shadow_mat4x3h: `enable f16;\nalias mat4x3h = mat4x3<f16>;`,
+  shadow_mat4x4h: `enable f16;\nalias mat4x4h = mat4x4<f16>;`,
```

Comment:
Maybe also test shadowing the unparameterized `mat2x2` forms?
e.g. `alias mat2x2 = array<vec2f, 2>;`

Reply:
Done in latest commit.